File: client/cli/src/commands/transactions.ts
import type { Argv, CommandBuilder, CommandModule } from 'yargs';
import { loadBudget } from '@tally/lib/data/loader';
import { Account } from '@tally/lib/core/account';
import { Budget } from '@tally/lib/core/budget';
import { Month } from '@tally/lib/core/month';
import { Transaction } from '@tally/lib/statement/transaction';
import {
  TransactionStatement,
  buildTransactionStatementTable,
} from '@tally/lib/statement/transaction';

type Options = {
  owner: string | undefined;
  account: string | undefined;
  startMonth: Month | undefined;
  endMonth: Month | undefined;
  limit: number | undefined;
};

type UnaccountedEntry = {
  account: Account;
  statement: TransactionStatement;
};

const command: string = 'transactions';
const desc: string = 'List of transactions within given period.';

const builder: CommandBuilder<unknown, Options> = (yargs: Argv<unknown>) =>
  yargs
    .options({
      owner: { type: 'string' },
      'start-month': { type: 'string' },
      'end-month': { type: 'string' },
      limit: { type: 'number', default: 20 },
    })
    .positional('account', { type: 'string' })
    .coerce('start-month', Month.fromString)
    .coerce('end-month', Month.fromString) as unknown as Argv<Options>;

export const commandModule: CommandModule<unknown, Options> = {
  command,
  describe: desc,
  builder,
  handler: async ({ owner, account, startMonth, endMonth, limit }): Promise<void> => {
    const budget: Budget = (await loadBudget()).budget;
    const statementTable: TransactionStatement[] = buildTransactionStatementTable(budget);
    let entries = new Map<string, Transaction[]>();
    for (const transactionStatement of statementTable) {
      const stmtAccount: Account = transactionStatement.account;
      if (owner && !stmtAccount.owners.includes(owner)) {
        continue;
      }
      if (account && stmtAccount.name !== account) {
        continue;
      }
      if (startMonth && transactionStatement.month.isLess(startMonth)) {
        continue;
      }
      if (endMonth && endMonth.isLess(transactionStatement.month)) {
        continue;
      }
      let accountEntries = entries.get(transactionStatement.account.name);
      if (!accountEntries) {
        accountEntries = [];
        entries.set(transactionStatement.account.name, accountEntries);
      }
      accountEntries.push(...transactionStatement.transactions);
    }
    for (const [, accountEntries] of entries) {
      accountEntries.sort((a, b) => {
        const timeDiff = a.balance.date.getTime() - b.balance.date.getTime();
        if (timeDiff !== 0) {
          return timeDiff;
        }
        return Math.abs(a.balance.amount) - Math.abs(b.balance.amount);
      });
    }
    process.stdout.write('Date,Amount,From,To,Description\n');
    for (const [accountName, accountEntries] of entries) {
      for (const t of accountEntries.slice(0, limit)) {
        let amount = t.balance.amount;
        const fromAccount = (amount < 0) ? t.account.name : accountName;
        const toAccount = (amount < 0) ? accountName : t.account.name;
        if (amount < 0) amount = -amount;
        process.stdout.write(
          `${t.balance.date.toISOString().slice(0, 10)},${
            (amount / 100).toFixed(2).padStart(8)}, ${
              fromAccount.padEnd(20)},${toAccount},${t.description ?? ''}\n`
        );
      }
    }
    process.exit(0);
  },
};

File: server/kt/src/main/kotlin/com/cherba29/tally/cli/cmds/Transactions.kt
package com.cherba29.tally.cli.cmds

import com.cherba29.tally.core.Month
import com.cherba29.tally.data.Loader
import com.cherba29.tally.statement.Transaction
import com.cherba29.tally.statement.TransactionStatement
import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.Context
import com.github.ajalt.clikt.parameters.arguments.argument
import com.github.ajalt.clikt.parameters.arguments.optional
import com.github.ajalt.clikt.parameters.options.convert
import com.github.ajalt.clikt.parameters.options.default
import com.github.ajalt.clikt.parameters.options.option
import com.github.ajalt.clikt.parameters.options.required
import com.github.ajalt.clikt.parameters.types.int
import com.github.ajalt.clikt.parameters.types.path
import kotlin.math.abs

class Transactions : CliktCommand() {
  override fun help(context: Context) = "List of transactions within given period."

  val tallyPath by option(envvar = "TALLY_PATH").path(mustExist = true).required()

  val account by argument(help = "Account name").optional()
  val startMonth by option(
    "--start-month",
    help="Inclusive start month, eg Apr2026"
  ).convert { Month.fromString(it) }
  val endMonth by option(
    "--end-month",
    help="Inclusive end month, eg May2026"
  ).convert { Month.fromString(it) }
  val owner: String? by option(
    "--owner",
    help="Owner of the account"
  )
  val limit: Int by option(
    "--limit",
    help="Number of transactions to list"
  ).int().default(20)

  override fun run() {
    val budget = Loader.loadFrom(tallyPath)
    val entries = mutableMapOf<String, MutableList<Transaction>>()
    for ((treeNode, monthTransactionStatements) in budget.nodeToStatement) {
      if (treeNode.children.isNotEmpty()) continue  // Only leaf nodes get processed.
      if (owner != null && owner != treeNode.path.first()) {
        continue
      }
      if (account != null && treeNode.name != account) {
        continue
      }
      for (transactionStatement in monthTransactionStatements.values) {
        if (startMonth != null && transactionStatement.monthRange.last < startMonth!!) {
          continue
        }
        if (endMonth != null && endMonth!! < transactionStatement.monthRange.first) {
          continue
        }
        var accountEntries = entries[transactionStatement.treeNode.name]
        if (accountEntries == null) {
          accountEntries = mutableListOf()
          entries[transactionStatement.treeNode.name] = accountEntries
        }
        accountEntries += (transactionStatement as TransactionStatement).transactions
      }
    }
    for (accountEntries in entries.values) {
      accountEntries.sortWith(Comparator { a, b ->
        if (a.balance.date == b.balance.date) {
          abs(a.balance.amount).compareTo(abs(b.balance.amount))
        } else if (a.balance.date < b.balance.date) -1 else 1
      })
    }
    echo("Date,Amount,From,To,Description")
    for ((accountName, accountEntries) in entries) {
      for (t in accountEntries.slice(0..limit)) {
        val amount = t.balance.amount
        echo(
          "${t.balance.date},${amount.asAmount().padStart(8)}, ${
            accountName.padEnd(20)
          },${t.treeNode.name},${t.description ?: ""}\n"
        )
      }
    }
  }

  companion object {
    private fun Long.asAmount(): String = "%.2f".format(this / 100.0)
  }
}